lib/dockerregistry/checks.go

```diff
@@ -245,3 +253,193 @@ func (check *ImageSizeCheck) Run() error {
+			}
+			// Only check the vulnerability for the source image if it
+			// hasn't been checked already since multiple promotion
+			// edges can contain the same source image
+			if _, found := srcImages[srcImage]; found {
+				continue
+			}
+			srcImages[srcImage] = nil
+			var req stream.ExternalRequest
+			req.RequestParams = edge
+			wg.Add(1)
+			reqs <- req
+		}
+	}
+
+	// If no custom ImageVulnProducer is provided, we use the default producer
+	// which is simply a call to the Container Analysis API which lists out
+	// the vulnerability occurrences for a given image
+	var vulnProducer ImageVulnProducer
+	if check.FakeVulnProducer != nil {
+		vulnProducer = check.FakeVulnProducer
+	} else {
+		ctx := context.Background()
+		client, err := containeranalysis.NewClient(ctx)
+		if err != nil {
+			return fmt.Errorf("NewClient: %v", err)
+		}
+		defer client.Close()
+		vulnProducer = mkRealVulnProducer(client)
+	}
+
+	vulnerableImages := make([]string, 0)
+	var processRequest ProcessRequest = func(
+		sc *SyncContext,
+		reqs chan stream.ExternalRequest,
+		requestResults chan<- RequestResult,
+		wg *sync.WaitGroup,
+		mutex *sync.Mutex) {
+		for req := range reqs {
+			reqRes := RequestResult{Context: req}
+			errors := make(Errors, 0)
+			edge := req.RequestParams.(PromotionEdge)
+			occurrences, err := vulnProducer(edge)
+			if err != nil {
+				errors = append(errors, Error{
+					Context: "error getting vulnerabilities",
+					Error:   err})
+			}
+			if len(occurrences) > 0 {
+				severeOccurrences := 0
+				for _, occ := range occurrences {
+					if CheckSeverity(occ, check.SeverityThreshold) {
+						errors = append(errors, Error{
+							Context: "vulnerabilities severity check",
+							Error: ImageVulnError{
+								edge.SrcImageTag.ImageName,
+								edge.Digest,
+								occ,
+							},
+						})
+						severeOccurrences++
+					}
+				}
+				if severeOccurrences > 0 {
+					vulnerableImages = append(vulnerableImages,
+						fmt.Sprintf("%v@%v [%v severe vulnerabilities, %v total]",
+							edge.SrcImageTag.ImageName,
+							edge.Digest,
+							severeOccurrences,
+							len(occurrences)))
+				}
+			}
+			reqRes.Errors = errors
+			requestResults <- reqRes
+		}
+	}
+
+	err := check.SyncContext.ExecRequests(
+		populateRequests,
+		processRequest,
+	)
+	if err != nil {
+		sort.Strings(vulnerableImages)
+		return fmt.Errorf("VulnerabilityCheck: "+
+			"The following vulnerable images were found:\n    %v",
+			strings.Join(vulnerableImages, "\n    "))
+	}
+	return nil
+}
+
+// Error is a function of ImageSizeError and implements the error interface.
+func (err ImageVulnError) Error() string {
+	vulnJSON, _ := json.MarshalIndent(err.Vulnerability, "", "  ")
+	return fmt.Sprintf("The image %v@%v contains the following "+
+		"vulnerability:\n%v",
+		err.ImageName, err.Digest, string(vulnJSON))
+}
+
+// CheckSeverity checks if a vulnerability is a high enough severity to fail
+// the ImageVulnCheck.
+func CheckSeverity(
```

Comment:
Because this function reduces down to a boolean check, I think a better name is`SevereOccurrence`. That way you can use language like `if SevereOccurrence(...) { ... }` in the codebase. WDYT?

Reply:
maybe IsSevereOccurrence so I can say if IsSevereOccurrence(...)

Reply:
Done